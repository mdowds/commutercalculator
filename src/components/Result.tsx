import * as React from "react";
import { JourneyResult, SeasonTickets, Travelcard } from "../types";
import { CSSProperties } from "react";

interface ResultProps {
  readonly result: JourneyResult;
  readonly isSelected: boolean;
  readonly averageHousePrice?: number;
  onSelectResult(selectedResult: JourneyResult): void;
}

export default class Result extends React.Component<ResultProps, {}> {
  travelcardDescription(travelcard: Travelcard) {
    return (
      <div>
        Zone {travelcard.minZone} to {travelcard.maxZone} travelcard: £
        {travelcard.price}
      </div>
    );
  }

  housePriceDescription() {
    const outcode = this.props.result.origin.postcode.split(" ")[0];
    const price = this.props.averageHousePrice
      ? "£" + this.props.averageHousePrice.toLocaleString("en-GB")
      : "Loading";
    return (
      <div>
        Average house price for {outcode}: {price}
      </div>
    );
  }

  makeDetails(result: JourneyResult) {
    const travelcardDiv = result.seasonTickets.travelcard
      ? this.travelcardDescription(result.seasonTickets.travelcard)
      : null;
    const seasonTicketDiv = result.seasonTickets.seasonTicket ? (
      <div>
        Direct season ticket: £{result.seasonTickets.seasonTicket.price}
      </div>
    ) : null;

    const seasonTicketsDescription = this.seasonTicketsHasValues(
      result.seasonTickets
    ) ? (
      <div>
        {travelcardDiv}
        {seasonTicketDiv}
      </div>
    ) : (
      <div>Season ticket not found</div>
    );

    const directionsLink = (
      <span>
        <a href={result.directionsUrl}>View directions on Google Maps</a> (May
        show a different journey time)
      </span>
    );

    return (
      <div>
        {seasonTicketsDescription}
        {this.housePriceDescription()}
        {directionsLink}
      </div>
    );
  }

  seasonTicketsHasValues(seasonTickets: SeasonTickets) {
    return seasonTickets.travelcard || seasonTickets.seasonTicket;
  }

  render() {
    const bgColour = this.props.isSelected ? "#ddd" : "white";

    const divStyle: CSSProperties = {
      padding: 10,
      borderBottom: "1px solid grey",
      backgroundColor: bgColour
    };

    const details = this.props.isSelected ? (
      <div style={{ margin: "10px 0px 0px 30px", fontSize: "75%" }}>
        {this.makeDetails(this.props.result)}
      </div>
    ) : null;

    const icon = this.props.isSelected
      ? "img/chevron-up.svg"
      : "img/chevron-down.svg";

    return (
      <div
        style={divStyle}
        onClick={() => {
          this.props.onSelectResult(this.props.result);
        }}
      >
        <div>
          <div style={{ float: "left", marginRight: 10 }}>
            <img
              alt="Expand results"
              src={icon}
              style={{ width: 20, height: 20 }}
            />
          </div>
          <div style={{ display: "inline-block", width: "65%" }}>
            {this.props.result.origin.name}
          </div>
          <div style={{ float: "right" }}>
            {this.props.result.journeyTime} mins
          </div>
        </div>
        {details}
      </div>
    );
  }
}
