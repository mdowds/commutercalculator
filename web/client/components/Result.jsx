"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
function travelcardDescription(travelcard) {
    return <div>Zone {travelcard.minZone} to {travelcard.maxZone} travelcard: £{travelcard.price}</div>;
}
function makeDetails(result) {
    var travelcardDiv = result.seasonTickets.travelcard ? travelcardDescription(result.seasonTickets.travelcard) : null;
    var seasonTicketDiv = result.seasonTickets.seasonTicket ? <div>Direct season ticket: £{result.seasonTickets.seasonTicket.price}</div> : null;
    var seasonTicketsDescription = <div>
        {travelcardDiv}
        {seasonTicketDiv}
    </div>;
    return seasonTicketsHasValues(result.seasonTickets) ?
        seasonTicketsDescription :
        <div>Season ticket not found</div>;
}
function seasonTicketsHasValues(seasonTickets) {
    return seasonTickets.travelcard || seasonTickets.seasonTicket;
}
function Result(props) {
    var bgColour = props.isSelected ? '#ddd' : 'white';
    var divStyle = {
        padding: 10,
        borderBottom: "1px solid grey",
        backgroundColor: bgColour
    };
    var details = props.isSelected ? <div style={{ margin: '10px 0px 0px 30px', fontSize: '75%' }}>{makeDetails(props.result)}</div> : null;
    var icon = props.isSelected ? "img/chevron-up.svg" : "img/chevron-down.svg";
    return (<div style={divStyle} onClick={function () { props.onSelectResult(props.result); }}>
            <div>
                <div style={{ float: "left", marginRight: 10 }}><img src={icon} style={{ width: 20, height: 20 }}/></div>
                <div style={{ display: "inline-block", width: "65%" }}>{props.result.origin.name}</div>
                <div style={{ float: "right" }}>{props.result.journeyTime} mins</div>
            </div>
            {details}
        </div>);
}
exports.default = Result;
