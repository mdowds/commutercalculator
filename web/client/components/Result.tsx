import * as React from 'react';
import {JourneyResult, Station} from '../types';
import {CSSProperties} from "react";

interface ResultProps {
    readonly result: JourneyResult,
    readonly isSelected: boolean,
    onSelectResult(selectedResult: string): void
}

function makeDetails(result: JourneyResult) {
    return result.seasonTicket ? <div>Season ticket price: Â£{result.seasonTicket.price}</div> : null;
}

export default function Result(props: ResultProps) {

    const bgColour = props.isSelected ? '#ddd' : 'white';

    const divStyle: CSSProperties = {
        padding: 10,
        borderBottom: "1px solid grey",
        backgroundColor: bgColour
    };

    const details = props.isSelected ? <div style={{margin: '10px 0px 0px 30px', fontSize: '75%'}}>{makeDetails(props.result)}</div> : null;

    const icon = props.isSelected ? "img/chevron-up.svg" : "img/chevron-down.svg";

    return (
        <div style={divStyle} onClick={() => {props.onSelectResult(props.result.origin.id)}}>
            <div>
                <div style={{float: "left", marginRight: 10}}><img src={icon} style={{width: 20, height: 20}}/></div>
                <div style={{display: "inline-block", width: "65%"}}>{props.result.origin.name}</div>
                <div style={{float: "right"}}>{props.result.journeyTime} mins</div>
            </div>
            {details}
        </div>
    );
}
