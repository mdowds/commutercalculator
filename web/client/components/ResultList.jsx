import React from 'react';
import PropTypes from 'prop-types';
import Result from './Result.jsx'

export default function ResultList(props) {

    const entries = props.results.map((result) => {
        return <Result key={result.origin.id} origin={result.origin} journeyTime={result.journeyTime} />
    });

    const containerStyle = {
        width: "100%",
        background: "white",
        height: "50%",
        overflow: "scroll"
    };

    const loadingIndicator = <div style={{textAlign: "center", paddingTop: 10}}>Results loading</div>;

    return (
        <div style={containerStyle}>
            {props.isLoading ? loadingIndicator : entries}
        </div>
    );
}

ResultList.propTypes = {
    results: PropTypes.arrayOf(PropTypes.object),
    isLoading: PropTypes.bool
};
