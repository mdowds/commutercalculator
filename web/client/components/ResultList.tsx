import * as React from 'react';
import {default as Result} from './Result';
import {JourneyResult} from '../types';
import {CSSProperties} from "react";

interface ResultListProps {
    readonly results: Array<JourneyResult>,
    selectedResult?: JourneyResult,
    readonly styles?: CSSProperties,
    readonly isLoading: boolean,
    onSelectResult(selectedResult: JourneyResult): void;
}

export default function(props) {
    const entries = props.results.map((result) => {
        const showDetails = props.selectedResult == result;

        return <Result key={result.origin.id} result={result} isSelected={showDetails} onSelectResult={props.onSelectResult} />
    });

    const containerStyleDefault: CSSProperties = {
        width: "100%",
        background: "white",
        height: "100%",
        overflow: "scroll"
    };

    const containerStyle = props.styles === undefined ? containerStyleDefault : Object.assign(containerStyleDefault, props.styles);

    const loadingIndicator = <div style={{textAlign: "center", marginTop: 50}}>Results loading</div>;

    return (
        <div style={containerStyle}>
            {props.isLoading ? loadingIndicator : entries}
        </div>
    );
}
