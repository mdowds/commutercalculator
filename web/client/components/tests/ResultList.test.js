import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import ResultList from '../ResultList.jsx';
import Result from '../Result.jsx';

test("ResultList renders the container", () => {
    const list = renderer.create(<ResultList results={[]} isLoading={false} />);
    expect(list.toJSON()).toMatchSnapshot();
});

test("ResultList renders the loading indicator", () => {
    const list = renderer.create(<ResultList results={[]} isLoading={true} />);
    expect(list.toJSON()).toMatchSnapshot();
});

test("ResultList renders results", () => {
    const list = shallow(<ResultList results={[mockResult(1, 10), mockResult(2, 20)]} isLoading={false} />);
    expect(list.find(Result).length).toEqual(2);
});

function mockResult(id, time) {
    return {
        origin: {id: id},
        journeyTime: time
    }
}
