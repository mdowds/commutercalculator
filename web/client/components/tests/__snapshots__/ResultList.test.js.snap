// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResultList renders the container 1`] = `
<div
  style={
    Object {
      "background": "white",
      "height": "50%",
      "overflow": "scroll",
      "width": "100%",
    }
  }
/>
`;

exports[`ResultList renders the loading indicator 1`] = `
<div
  style={
    Object {
      "background": "white",
      "height": "50%",
      "overflow": "scroll",
      "width": "100%",
    }
  }
>
  <div
    style={
      Object {
        "paddingTop": 10,
        "textAlign": "center",
      }
    }
  >
    Results loading
  </div>
</div>
`;
